### DISCUSSION TITLE
Outline Rejects Microsoft Authentication

### DISCUSSION THREAD

**Discussion by mjducut-oe** [outline/outline#7544] [Self Hosting] [answered]
Hi there! 

The requirement is to have Microsoft as the main authentication to be able to use our outline instance. I have followed how to set it up using the documentation and did not have any luck. Also tried referring to previous issues but no luck there either. I'm pretty sure the app registration is working because I can see a successful logins using the sign-in logs from Entra. 

I have tried integrating discord authentication and this one seems to work. We appreciate if anyone can help us in this regard. 

We are nearly finished with this one - only need to fix the Microsoft Authentication. 

**To Reproduce**
1. Go to to main login page.
2. Click on "Login with Microsot"
3. This is the result after the Microsoft Authentication: 
![image](https://github.com/user-attachments/assets/852c8509-0870-4267-833f-71e5765c1802)

**Expected behavior**
To be able to login properly using Microsoft Authentication as our app registration seems to be properly setup. 

**Screenshots**
Right off the bat, when I go to our login page, this error can be immidiately seen from the browser's developer console even before I click the login buttons: 
![image](https://github.com/user-attachments/assets/8e7f3d1d-d750-4666-b937-db38584ab5bd)

When I login with Microsoft, this was the container logs: 
![image](https://github.com/user-attachments/assets/900939b8-3dcd-4bc5-a1d9-7e8c0dcba0f8)

Here's my app registration api permissions. Even with openId included, this still doesn't work:
![image](https://github.com/user-attachments/assets/b2cf2f32-174d-4b23-b1b9-f1632c94ea6f)

**Outline (please complete the following information):**
- Install: Self-hosted 
- Version: 0.79.0

**Desktop (please complete the following information):**
 - OS: Windows
 - Browser: Chrome
 - Version: 128.0.6613.113

**Comment by tommoor** [comment]
Try adding `Organization.Read.All`, it shouldn't be needed but the error is indicative of not being able to query org details

**Reply by mjducut-oe** [reply]
Thanks for the response. I've tried adding that permission to but I'm still getting the same issue as described above. 

![image](https://github.com/user-attachments/assets/c7a346c1-5de0-4d45-87cf-0c3335a86c84)


Not sure if this will help but every time I click on Continue with Microsoft and got redirected to the MS login page, here's the URL that's being used:

`https://login.microsoftonline.com/xxxx-xxxxx-xxxx-xxxxxx/oauth2/authorize?prompt=select_account&originalQuery=&response_type=code&redirect_uri=https%3A%2F%2Fmydomain.com%2Fauth%2Fazure.callback&scope=&state=xxxxxxxxxxx&client_id=xxxxxxxxxx`

Not sure if that's properly formatted and if `scope` should have any value?

**Reply by mjducut-oe** [reply]
Also just to add, could this be an issue? Can see that I am receiving the authorization code but for some reason, it's prefixed with a `0.`? 
![image](https://github.com/user-attachments/assets/da5af485-9715-4d61-86e0-d4ada481e32c)

**Reply by tommoor** [reply]
That all seems fine. Change your logging level to the most verbose `LOG_LEVEL=silly` and try again you'll see the full output of the request and response to Microsoft which should provide everything we need to know what's going on.

**Reply by mjducut-oe** [reply]
Hi @tommoor, thanks for the assistance. 

I've changed LOG_LEVEL to silly and this is still what I was getting:
![image](https://github.com/user-attachments/assets/0da9aa5e-9111-4503-9a59-fa75294447f6)

**Reply by tommoor** [reply]
Did you upload the correct image? There is no new logs here

**Reply by PowellTravis** [reply]
Good Evening I am encountering this same issue using Microsoft Entra for authentication. I am seeing the same result as the original poster on the Entra side as successful logins. Set LOG_LEVEL=silly

``` log
Oct 16 01:36:51 docs outline[11910]: {"label":"http","level":"info","message":"  <-- GET /auth/azure.callback?code=<Entra Callback Code>&state=<State>&session_state=<Session State>"}
Oct 16 01:36:52 docs outline[11910]: {"error":"Cannot read properties of undefined (reading '0')","level":"error","message":"Error during authentication","stack":"TypeError: Cannot read properties of undefined (reading '0')\n    at Strategy._verify (/var/outline/build/plugins/azure/server/auth/azure.js:52:54)\n    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)"}
Oct 16 01:36:52 docs outline[11910]: {"label":"http","level":"info","message":"  --> GET /auth/azure.callback?code=<Entra Callback Data>"}
```

**Reply by tommoor** [reply]
It means that the request to `https://graph.microsoft.com/v1.0/organization` returned no value. (https://docs.microsoft.com/en-us/graph/api/organization-get?view=graph-rest-1.0), It's likely to be a permission / scope error of some sort.

**Reply by PowellTravis** [reply]
The permissions I have delegated to the application are below. I don't think I am missing anything regarding the permissions.
<img width="500" alt="Screenshot 2024-10-16 at 9 49 46 AM" src="https://github.com/user-attachments/assets/cc679d3c-5a54-419d-8935-135137939a3a">

**Reply by rsteckly** [reply]
Same issue, following.

**Reply by tommoor** [reply]
I just published an image tagged `0.80.2-ms-debug` which includes logging of the response from this API endpoint so we can see what's going on, I'll need someone to switch to that and post the resulting logs (you can obfuscate ID's, it's really the general shape that matters)

**Reply by PowellTravis** [reply]
I updated and switched my installation to 0.80.2-ms-debug and got the following output.
``` log
Oct 17 01:35:59 docs outline[27409]: {"label":"http","level":"info","message":"  <-- GET /auth/azure.callback?code=code=<session code>&state=<State>&session_state=<session state>"}
Oct 17 01:35:59 docs outline[27409]: {"error":{"code":"InvalidAuthenticationToken","innerError":{"client-request-id":"<client_request_id>","date":"2024-10-17T01:35:59","request-id":"<request_id>"},"message":"[Filtered]"},"label":"authentication","level":"info","message":"API response missing value key"}
Oct 17 01:35:59 docs outline[27409]: {"error":"Unable to load organization value from Microsoft Graph API","level":"error","message":"Error during authentication","stack":"BadRequestError: Unable to load organization value from Microsoft Graph API\n    at MicrosoftGraphError (/var/outline/build/server/errors.js:167:34)\n    at Strategy._verify (/var/outline/build/plugins/azure/server/auth/azure.js:55:47)\n    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)"}
Oct 17 01:35:59 docs outline[27409]: {"label":"http","level":"info","message":"  --> GET /auth/azure.callback?code=<session code>&state=<State>&session_state=<session state>"
```

I hope this helps

**Reply by tommoor** [reply]
It gets us one step further, unfortunately the message was replaced with `[Filtered]` as it contained the word "token", so will probably need another build to reveal that

**Reply by rsteckly** [reply]
Do you think its that we need to grant more permissions?  I mean getoutline the hosted version works...so it is working in some version of the app.  I wonder if we need to grant application rather than delegated permissions?

**Reply by tommoor** [reply]
Delegated is correct, as documented. But yes, it's a permissions error of _some_ sort.

**Reply by rsteckly** [reply]
I went through and added nearly all the delegated permissions (I still have 10 categories to try).  Still getting the same error.  We don't have a lot protected by Entra, so I am thinking I might just give it all delegated permissions to see if it that resolves it.  Then we can take away section by section.

**Reply by tommoor** [reply]
The build tagged `0.80.2-ms-debug-2` should include the `[Filtered]` message in the logs now.

**Reply by PowellTravis** [reply]
Here is the output when on 0.80.2-ms-debug-2

``` log
Oct 18 16:33:50 docs outline[48737]: {"label":"http","level":"info","message":"  <-- GET /auth/azure.callback?code=<code>&state=<state>&session_state=<session_state>"}
Oct 18 16:33:50 docs outline[48737]: {"error":{"code":"InvalidAuthenticationToken","innerError":{"client-request-id":"<request-id>","date":"2024-10-18T16:33:50","request-id":"<request-id>"},"message":"Access token validation failure. Invalid audience."},"label":"authentication","level":"info","message":"API response missing value key"}
Oct 18 16:33:50 docs outline[48737]: {"error":"Unable to load organization value from Microsoft Graph API","level":"error","message":"Error during authentication","stack":"BadRequestError: Unable to load organization value from Microsoft Graph API\n    at MicrosoftGraphError (/var/outline/build/server/errors.js:167:34)\n    at Strategy._verify (/var/outline/build/plugins/azure/server/auth/azure.js:55:47)\n    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)"}
Oct 18 16:33:50 docs outline[48737]: {"label":"http","level":"info","message":"  --> GET /auth/azure.callback?code=<code>&state=<state>&session_state=<session_state>"}
```

**Reply by tommoor** [reply]
What value do you have set in the environment for AZURE_RESOURCE_ID?

**Reply by PowellTravis** [reply]
I have 00000003-0000-0000-c000-000000000000 set as the AZURE_RESOURCE_ID

**Reply by tommoor** [reply]
I'd be interested about others in this thread, an incorrect resource ID seems to be the most likely culprit for this error.

https://stackoverflow.com/questions/58428426/fetching-profile-image-from-azure-using-passport-token

**Reply by PowellTravis** [reply, marked as answer]
@tommoor I got it working. The error was I had AZURE_RESOURCE_ID and the plugin/azure/server/env.ts file was looking for AZURE_RESOURCE_APP_ID and with my config set as:
``` env
AZURE_CLIENT_ID="<Application (Client) ID>"
AZURE_CLIENT_SECRET="<App > Certificates & Secrets > Client Secret Value"
AZURE_TENANT_ID="<Directory (tenant) ID>"
AZURE_RESOURCE_APP_ID="00000003-0000-0000-c000-000000000000"
```
Entra authenticates with the following API Permissions exposed

![Screenshot 2024-10-18 at 10 22 18 PM](https://github.com/user-attachments/assets/fccffb2f-f1e4-4952-b1be-8c8deb268dd3)

``` ts
class AzurePluginEnvironment extends Environment {
  /**
   * Azure OAuth2 client credentials. To enable authentication with Azure.
   */
  @IsOptional()
  @CannotUseWithout("AZURE_CLIENT_SECRET")
  public AZURE_CLIENT_ID = this.toOptionalString(environment.AZURE_CLIENT_ID);

  @IsOptional()
  @CannotUseWithout("AZURE_CLIENT_ID")
  public AZURE_CLIENT_SECRET = this.toOptionalString(
    environment.AZURE_CLIENT_SECRET
  );

  @IsOptional()
  @CannotUseWithout("AZURE_CLIENT_ID")
  public AZURE_RESOURCE_APP_ID = this.toOptionalString(
    environment.AZURE_RESOURCE_APP_ID
  );

  @IsOptional()
  @CannotUseWithout("AZURE_CLIENT_ID")
  public AZURE_TENANT_ID = this.toOptionalString(environment.AZURE_TENANT_ID);
}
```

**Reply by tommoor** [reply]
Landed two changes for the next version,

- `AZURE_RESOURCE_APP_ID` is now set with a default value
- Improved error handling for these requests to MS

**Reply by rsteckly** [reply]
This did not resolve the issue for us.  We added tenant id....but we are still getting the same error.  We have verified all the values are correct.

**Reply by PowellTravis** [reply]
What does your azure config look like, with values redacted?

**Reply by tommoor** [reply]
If it didn't work then post your logs from the last debug build

**Reply by hhaswan** [reply]
I tried with this permission. and login successfully ✅ 
![image](https://github.com/user-attachments/assets/f589a59d-5ebf-43f5-bf46-826d0aa76e2c)

and use this variable : 
AZURE_CLIENT_ID=
AZURE_CLIENT_SECRET=
AZURE_TENANT_ID=

**Comment by rsteckly** [comment]
We are having the same issue.  We tried adding the organization..read permissions and it didn't help.  I don't think we are going to be able to use Outline unless we can get Office 365 working.  This has stopped us in our tracks.

**Comment by routerino** [comment]
Perhaps you could enroll Microsoft as a generic OIDC app instead of using the oauth integration?
